data/transactions/logic/resources.go

```diff
@@ -0,0 +1,359 @@
+	id := tx.XferAsset
+	r.sharedAsas[id] = struct{}{}
+	r.shareAccountAndHolding(hdr.Sender, id)
+	r.shareAccountAndHolding(tx.AssetReceiver, id)
+
+	if !tx.AssetSender.IsZero() {
+		r.shareAccountAndHolding(tx.AssetSender, id)
+	}
+
+	if !tx.AssetCloseTo.IsZero() {
+		r.shareAccountAndHolding(tx.AssetCloseTo, id)
+	}
+}
+
+// allowsHolding checks if a holding is available under the txgroup sharing rules
+func (cx *EvalContext) allowsHolding(addr basics.Address, ai basics.AssetIndex) bool {
+	r := cx.available
+	if _, ok := r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: ai}]; ok {
+		return true
+	}
+	// If an ASA was created in this group, then allow holding access for any allowed account.
+	for _, created := range r.createdAsas {
+		if created == ai {
+			return cx.availableAccount(addr)
+		}
+	}
+	// If the address was "created" by making its app in this group, then allow for available assets.
+	for _, created := range r.createdApps {
+		if cx.getApplicationAddress(created) == addr {
+			return cx.availableAsset(ai)
+		}
+	}
+	// If the current txn is a creation, the new appID won't be in r.createdApps
+	// yet, but it should get the same special treatment.
+	if cx.txn.Txn.ApplicationID == 0 && cx.getApplicationAddress(cx.appID) == addr {
+		return cx.availableAsset(ai)
+	}
+	return false
+}
+
+// allowsLocals checks if a local state is available under the txgroup sharing rules
+func (cx *EvalContext) allowsLocals(addr basics.Address, ai basics.AppIndex) bool {
+	r := cx.available
+	if _, ok := r.sharedLocals[ledgercore.AccountApp{Address: addr, App: ai}]; ok {
+		return true
+	}
+	// All locals of created apps are available
+	for _, created := range r.createdApps {
+		if created == ai {
+			return cx.availableAccount(addr)
+		}
+	}
+	if cx.txn.Txn.ApplicationID == 0 && cx.appID == ai {
+		return cx.availableAccount(addr)
+	}
+
+	// All locals of created app accounts are available
+	for _, created := range r.createdApps {
+		if cx.getApplicationAddress(created) == addr {
+			return cx.availableApp(ai)
+		}
+	}
+	if cx.txn.Txn.ApplicationID == 0 && cx.getApplicationAddress(cx.appID) == addr {
+		return cx.availableApp(ai)
+	}
+	return false
+}
+
+func (cx *EvalContext) requireHolding(acct basics.Address, id basics.AssetIndex) error {
+	/* Previous versions allowed inner appls with zeros in "required" places,
+	   even if that 0 resource should have be inaccessible, because the check
+	   was done at itxn_field time, and maybe the app simply didn't set the
+	   field. */
+	if id == 0 || acct.IsZero() {
+		return nil
+	}
+	if !cx.allowsHolding(acct, id) {
+		return fmt.Errorf("invalid Holding access %s x %d would be possible", acct, id)
+	}
+	return nil
+}
+
+func (cx *EvalContext) requireLocals(acct basics.Address, id basics.AppIndex) error {
+	if !cx.allowsLocals(acct, id) {
+		return fmt.Errorf("invalid Local State access %s x %d would be possible", acct, id)
```

Comment:
interesting so an assettransfer sender has to be opted into the asset even if it the assetsender is unique.  Is this for clawback?

Reply:
I meant transactions where AssetSender is not the same as Sender